nutils/topology.py

```diff
@@ -87,9 +87,166 @@ def __str__(self) -> str:
+    the slice. A selection of the full slice is always possible. A selection of
+    smaller slices is possible for products of topologies, provided that the
+    product can be factorized in a (possible zero-dimensional) topology before
+    the slice, the slice and a (possibly zero-dimensional) topology after the
+    slice.
+
+    Parameters
+    ----------
+    indices : integer :class:`numpy.ndarray` or similar
+        The one-dimensional array of element indices.
+    idim : :class:`int`
+        The begining of the slice of the dimension to operate on.
+    ndim : :class:`int`
+        The length of the slice of the dimension to operate on.
+
+    Returns
+    -------
+    :class:`Topology`
+        The selected elements.
+
+    See Also
+    --------
+    :meth:`compress` : select elements using a mask
+    '''
+
+    if not 0 <= __idim <= __idim + __ndim <= self.ndims:
+      raise IndexError('dimension indices out of range')
+    indices = numpy.asarray(__indices)
+    if indices.ndim != 1:
+      raise ValueError('expected a one-dimensional array')
+    if not indices.size:
+      return self.empty_like()
+    indices = indices.astype(int, casting='same_kind')
+    return self.take_unchecked(indices, __idim, __ndim)
+
+  def take_unchecked(self, __indices: numpy.ndarray, __idim: int, __ndim: int) -> 'Topology':
+    raise NotImplementedError
+
+  def compress(self, __mask: Union[numpy.ndarray, Sequence[bool]], __idim: int, __ndim: int) -> 'Topology':
+    '''Return the selected elements as a disconnected topology.
+
+    The selection is applied to slice `idim:ndim` of the dimension of this
+    topology, if possible. The indices refer to the raveled list of elements in
+    the slice. A selection of the full slice is always possible. A selection of
+    smaller slices is possible for products of topologies, provided that the
+    product can be factorized in a (possible zero-dimensional) topology before
+    the slice, the slice and a (possibly zero-dimensional) topology after the
+    slice.
+
+    Parameters
+    ----------
+    indices : boolean :class:`numpy.ndarray` or similar
+        The one-dimensional element mask.
+    idim : :class:`int`
+        The begining of the slice of the dimension to operate on.
+    ndim : :class:`int`
+        The length of the slice of the dimension to operate on.
+
+    Returns
+    -------
+    :class:`Topology`
+        The selected elements.
+
+    See Also
+    --------
+    :meth:`compress` : select elements by index
+    '''
+
+    if not 0 <= __idim <= __idim + __ndim <= self.ndims:
+      raise IndexError('dimension indices out of range')
+    mask = numpy.asarray(__mask)
+    if mask.ndim != 1:
+      raise ValueError('expected a one-dimensional array')
+    if __idim == 0 and __ndim == self.ndims:
+      if mask.size != len(self):
+        raise ValueError('expected a one-dimensional array with length {}'.format(len(self)))
+    return self.compress_unchecked(mask, __idim, __ndim)
```

Comment:
I wonder if compress merits its own method; shouldn't we just have `__getitem__` hand the nonzeros to take instead?

Reply:
In the latest version there still is a `compress` method, but there are no redefinitions.